storage/blockdevice/COMPONENT_QSPIF/UNITTESTS/test_QSPIFBlockDevice.cpp

```diff
@@ -0,0 +1,285 @@
+{
+    // The real implementation of this function queries
+    // the CR3NV register on S25FS512S.
+    test_sfdp_reader = sfdp_reader;
+    return 0;
+}
+
+size_t sfdp_detect_page_size(uint8_t *bptbl_ptr, size_t bptbl_size)
+{
+    return 0;
+}
+
+int sfdp_detect_erase_types_inst_and_size(uint8_t *bptbl_ptr, sfdp_hdr_info &sfdp_info)
+{
+    return 0;
+}
+
+int sfdp_find_addr_region(bd_addr_t offset, const sfdp_hdr_info &sfdp_info)
+{
+    return 0;
+}
+
+int sfdp_iterate_next_largest_erase_type(uint8_t bitfield,
+                                         bd_size_t size,
+                                         bd_addr_t offset,
+                                         int region,
+                                         const sfdp_smptbl_info &smptbl)
+{
+    return 0;
+}
+
+int sfdp_detect_device_density(uint8_t *bptbl_ptr, sfdp_bptbl_info &bptbl_info)
+{
+    return 0;
+}
+
+int sfdp_detect_addressability(uint8_t *bptbl_ptr, sfdp_bptbl_info &bptbl_info)
+{
+    return 0;
+}
+
+} // namespace mbed
+
+
+class TestQSPIFBlockDevice : public testing::Test {
+protected:
+    TestQSPIFBlockDevice()
+        : bd(PinName(0), PinName(1), PinName(2), PinName(3), PinName(4), PinName(5)) // fake pins
+    {
+    }
+
+    virtual void SetUp()
+    {
+        EXPECT_EQ(QSPIF_BD_ERROR_OK, bd.init());
+    }
+
+    virtual void TearDown()
+    {
+        EXPECT_EQ(QSPIF_BD_ERROR_OK, bd.deinit());
+    }
+
+    QSPIFBlockDevice bd;
+};
+
+/**
+ * Test that the quirk on the reported values of the CR1NV and CR3NV
+ * registers are applied when the flash chip is S25FS512S.
+ */
+TEST_F(TestQSPIFBlockDevice, TestS25FS512SQuirk)
+{
+    // Read the CR1NV register
+    uint8_t CR1NV;
+    EXPECT_EQ(QSPIF_BD_ERROR_OK, mbed::test_sfdp_reader(
+                  S25FS512S_CR1NV,
+                  mbed::SFDP_CMD_ADDR_SIZE_VARIABLE,
+                  0x65,
+                  mbed::SFDP_CMD_DUMMY_CYCLES_VARIABLE,
+                  &CR1NV,
+                  sizeof(CR1NV)));
+
+    // Read the CR3NV register
+    uint8_t CR3NV;
+    EXPECT_EQ(QSPIF_BD_ERROR_OK, mbed::test_sfdp_reader(
+                  S25FS512S_CR3NV,
+                  mbed::SFDP_CMD_ADDR_SIZE_VARIABLE,
+                  0x65,
+                  mbed::SFDP_CMD_DUMMY_CYCLES_VARIABLE,
+                  &CR3NV,
+                  sizeof(CR3NV)));
+
+    // The hardware value of CR3NV[1] is 0 but S25FS512S's SFDP table
+    // expects it to be 1.
+    EXPECT_EQ(0x2, CR3NV & 0x02);
+
+    // The factory default configuration is CR1NV[2] == 0 and CR3NV[3] == 0
+    // (eight 4KB sectors overlaying the start of the flash) but we treat
+    // the flash chip as if CR1NV[2] == 0 and CR3NV[3] == 1 (no overlaying 4KB
+    // sectors), as Mbed OS does not support this type of flash layout.
+    EXPECT_EQ(0x0, CR1NV & 0x4);
+    EXPECT_EQ(0x8, CR3NV & 0x8);
+}
```

Comment:
Do we have a test to check that the quirk is not applied when the chip is not S25FS512S?